packages/opentelemetry-basic-tracer/test/export/BatchSpanProcessor.test.ts

```diff
@@ -0,0 +1,126 @@
+    this.spansDataList.push(...spans);
+  }
+
+  shutdown(): void {
+    this.spansDataList = [];
+  }
+}
+
+function createSampledSpan(spanName: string): Span {
+  const tracer = new BasicTracer({
+    scopeManager: new NoopScopeManager(),
+    sampler: ALWAYS_SAMPLER,
+  });
+  const span = tracer.startSpan(spanName);
+  span.end();
+  return span as Span;
+}
+
+function createUnSampledSpan(spanName: string): Span {
+  const tracer = new BasicTracer({
+    scopeManager: new NoopScopeManager(),
+    sampler: NEVER_SAMPLER,
+    logger: new NoopLogger(),
+  });
+  const span = tracer.startSpan(spanName, { isRecordingEvents: false });
+  span.end();
+  return span as Span;
+}
+
+describe('BatchSpanProcessor', () => {
+  const name = 'span-name';
+  const defaultBufferConfig = {
+    bufferSize: 5,
+    bufferTimeout: 2000,
+  };
+  const exporter = new TestExporter();
+
+  describe('constructor', () => {
+    it('should create a BatchSpanProcessor instance', () => {
+      const processor = new BatchSpanProcessor(exporter);
+      assert.ok(processor instanceof BatchSpanProcessor);
+    });
+
+    it('should create a BatchSpanProcessor instance with config', () => {
+      const processor = new BatchSpanProcessor(exporter, defaultBufferConfig);
+      assert.ok(processor instanceof BatchSpanProcessor);
+    });
+
+    it('should create a BatchSpanProcessor instance with empty config', () => {
+      const processor = new BatchSpanProcessor(exporter, {});
+      assert.ok(processor instanceof BatchSpanProcessor);
+    });
+  });
+
+  describe('.onStart/.onEnd/.shutdown', () => {
+    it('should export the sampled spans with buffer size reached', () => {
+      const processor = new BatchSpanProcessor(exporter, defaultBufferConfig);
+      for (let i = 0; i < defaultBufferConfig.bufferSize; i++) {
+        const span = createSampledSpan(`${name}_${i}`);
+        processor.onStart(span);
+        assert.strictEqual(exporter.spansDataList.length, 0);
+
+        processor.onEnd(span);
+        assert.strictEqual(exporter.spansDataList.length, 0);
+      }
+      // Now we should start seeing the spans in exporter
+      const span = createSampledSpan(`${name}_6`);
+      processor.onEnd(span);
+      assert.strictEqual(exporter.spansDataList.length, 6);
+
+      processor.shutdown();
+      assert.strictEqual(exporter.spansDataList.length, 0);
+    });
+
+    it('should not export the unsampled spans', () => {
+      const processor = new BatchSpanProcessor(exporter, defaultBufferConfig);
+      for (let i = 0; i < defaultBufferConfig.bufferSize * 2; i++) {
+        const span = createUnSampledSpan(`${name}_${i}`);
+        processor.onEnd(span);
+        assert.strictEqual(exporter.spansDataList.length, 0);
+      }
+    });
+
+    it('should force flush when timeout exceeded', done => {
+      const processor = new BatchSpanProcessor(exporter, defaultBufferConfig);
+      for (let i = 0; i < defaultBufferConfig.bufferSize; i++) {
```

Comment:
Should we do something like `.bufferSize - 1` to explicitly show we are putting in less than the full buffer of spans and then seeing that it still gets flushed?